src/precice/impl/ParticipantImpl.cpp

```diff
@@ -1047,9 +1048,11 @@ void ParticipantImpl::readData(
     ::precice::span<double>         values) const
 {
   PRECICE_TRACE(meshName, dataName, vertices.size(), relativeReadTime);
+  PRECICE_CHECK(_state != State::Constructed, "readData(...) cannot be called before initialize().");
   PRECICE_CHECK(_state != State::Finalized, "readData(...) cannot be called after finalize().");
   PRECICE_CHECK(math::smallerEquals(relativeReadTime, _couplingScheme->getNextTimeStepMaxSize()), "readData(...) cannot sample data outside of current time window.");
   PRECICE_CHECK(relativeReadTime >= 0, "readData(...) cannot sample data before the current time.");
+  PRECICE_CHECK(isCouplingOngoing() || relativeReadTime == 0, "Calling readData(...) with relativeReadTime = {} is forbidden if coupling is not ongoing. If coupling finished, only relativeReadTime = 0 is allowed. Please always use precice.getMaxTimeStepSize() to obtain the maximum allowed relativeReadTime.", relativeReadTime);
```

Comment:
```suggestion
  PRECICE_CHECK(isCouplingOngoing() || relativeReadTime == 0, "Calling readData(...) with relativeReadTime = {} is forbidden if coupling is not ongoing. If coupling finished, only relativeReadTime = 0 is available. Please always use precice.getMaxTimeStepSize() to obtain the maximum allowed relativeReadTime.", relativeReadTime);
```

It's not something we came up with, but it's also that non-zero relativeReadTime is simply not available/ we are not able to compute it without extrapolating, right? It sounds at least for me quite different, whether we write 'we forbid' or 'what you are trying to do doesn't make sense/ we don't have the data you are requesting'.